core/src/main/java/org/infinispan/marshall/protostream/impl/MarshallableArray.java

```diff
@@ -0,0 +1,151 @@
+package org.infinispan.marshall.protostream.impl;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.infinispan.commons.marshall.MarshallingException;
+import org.infinispan.commons.marshall.ProtoStreamTypeIds;
+import org.infinispan.commons.util.Util;
+import org.infinispan.protostream.ProtobufTagMarshaller;
+import org.infinispan.protostream.TagReader;
+import org.infinispan.protostream.annotations.ProtoFactory;
+import org.infinispan.protostream.annotations.ProtoField;
+import org.infinispan.protostream.annotations.ProtoTypeId;
+import org.infinispan.protostream.annotations.impl.GeneratedMarshallerBase;
+import org.infinispan.protostream.descriptors.WireType;
+
+/**
+ * A wrapper for collections of objects whose type is unknown until runtime. This is equivalent to utilising a
+ * <code>Collection<MarshallableObject></code> without the overhead of creating a {@link MarshallableObject} per
+ * entry.
+ *
+ * @author Ryan Emerson
+ * @since 16.0
+ */
+@ProtoTypeId(ProtoStreamTypeIds.MARSHALLABLE_ARRAY)
+public class MarshallableArray<T> {
```

Comment:
Why do we need this class? 
Isn't protostream using this class to handle arrays of objects?
https://github.com/infinispan/protostream/blob/main/types/src/main/java/org/infinispan/protostream/types/java/arrays/ObjectArrayAdapter.java

Reply:
I forgot that small detail that users can configure a different marshaller :cry: 
Follow-up question, can the `collection` be removed? it is unnecessary if you send the array length.

Reply:
The idea of having both the `collection` and `array` field was to prevent a call to `toArray` unnecessarily. We could write the array length like you say, but then we're adding field ordering logic to the parser which we're technically not supposed to do:

https://protobuf.dev/programming-guides/encoding/#order
```quote
When a message is serialized, there is no guaranteed order for how its known or [unknown fields](https://protobuf.dev/programming-guides/proto2#updating) will be written. Serialization order is an implementation detail, and the details of any particular implementation may change in the future. Therefore, protocol buffer parsers must be able to parse fields in any order.
``` 

Given that this is a very niche use-case, we could make an exception, WDYT?

Reply:
Taking into consideration that we use `Arrays.hashCode` and `Arrays.equals` when fetching keys, the field order must be kept otherwise we are in big trouble.